File: tree/traverse/verticalTraversal.py
# level order, + distance to root (left is -1 distance, right is +1 distance)
# so leftest node alswasy has the smallest value/distance

# if same position, sort from left to right
def verticalOrder(self, root: TreeNode) -> List[List[int]]:
    if root is None:
        return []

    res = []
    queue = collections.deque([(0, root)])
    dist_dict = collections.defaultdict(list)

    while queue:
        node_dist, node = queue.popleft()
        dist_dict[node_dist].append(node.val)

        if node.left:
            queue.append((node_dist-1, node.left))
            # print(node.left.val)
        if node.right:
            queue.append((node_dist+1, node.right))
            # print(node.right.val)

    for distance, nodes in sorted(dist_dict.items()):
        res.append(nodes)

    return res

# if smae position, sort from small to big
#  record coordinate using extray array (x, y, node.val)
def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
    queue = collections.deque()
    queue.append((root, 0, 0))  # order does not matter, node, x, y
    arr = list()   # x, y, node.val
    while queue:
        node, x, y = queue.popleft()
        arr.append((x, y, node.val))
        if node.left:
            queue.append((node.left, x - 1, y + 1))
        if node.right:
            queue.append((node.right, x + 1, y + 1))

    arr.sort()

    res = [[arr[0][2]]]
    for i in range(1, len(arr)):
        if arr[i][0] == arr[i - 1][0]:  # smae distance in []
            res[-1].append(arr[i][2])
        else:
            res.append([arr[i][2]])
    return res

File: tree/traverse/verticalTraversal.java
//from left to rightS, if same vertical distance, then from top to bottom (level order)
//https://piazza.com/class/j0eqhhdregb3i?cid=2721
public List<List<Integer>> verticalOrder(TreeNode root) {
    List<List<Integer>> res = new LinkedList<>();
    if(root == null)
      return res;
    //<disToRoot, listOF(node_values)>
    Map<Integer, List<Integer>> map = new TreeMap<>();
    Queue<Item> queue = new LinkedList<>();
    queue.offer(new Item(0, root));
    //不需要先放左边的node进result，你依然可以从上到下的顺序遍历，use linkedlist offerlast and offerfirst
    while(!queue.isEmpty()){
      Item cur = queue.poll();
      TreeNode node = cur.node;
      //process res
      if(!map.contiansKey(cur.disToRoot)){
        map.put(cur.disToRoot, new ArrayList<>());
      }
      map.get(cur.disToRoot).add(node.key);  //level order for the nodes with the same distance

      if(node.left !=null)
        queue.offer(new Item(node.left, cur.disToRoot-1));
      if(node.right !=null)
        queue.offer(new Item(node.right, cur.disToRoot+1));
    }
    for(Map.Entry<Integer, List<Integer>> entry : map.entrySet()){
      res.add(entry.getValue());
    }
    return res;
  }

  static void printVertically(TreeNode root, TreeMap<integer, linkedlist<integer="">> result, int level) {
          if (root == null) {
              return;
          }

          LinkedList<integer> list = result.get(level);
          if (list == null) {
              list = new LinkedList<>();
          }
          list.add(root.data);
          result.put(level, list);

          printVertically(root.left, result, level-1);

          printVertically(root.right, result, level+1);
      }


//topView of binary TreeNode
//https://www.geeksforgeeks.org/print-nodes-top-view-binary-tree/
// A node x is there in output if x is the topmost node at its horizontal distance.
// Horizontal distance of left child of a node x is equal to horizontal distance of x minus 1,
// and that of right child is horizontal distance of x plus 1.
//level order: queue
    7
  6   5      => 7654
4  3

public List<TreeNode> topView(TreeNode root){
  List<TreeNode> res = new ArrayList<>();
  if(root == null)
    return res;
  Set<Integer> set = new HashSet<>();  //the shadow top
  Queue<Item> queue = new LinkedList<>();
  queue.offer(new Item(root, 0));

  while(!queue.isEmpty()){
    Item cur = queue.poll();
    TreeNode node = cur.node;
    // If this is the first node at its horizontal distance,then this node is in top view
    if(!set.contains(cur.disToRoot)){
      set.add(cur.disToRoot);
      res.add(cur);
    }

    if(node.left !=null)
      queue.offer(new Item(node.left, cur.disToRoot-1));
    if(node.right !=null)
      queue.offer(new Item(node.right, cur.disToRoot+1));
  }
}
static class Item{
  TreeNode node;
  int disToRoot;
  public Item(TreeNode node, int dis){
    this.node = node;
    this.disToRoot = dis;  //colum number
  }
}
